timemachine/cpp/src/neighborlist.cu

```diff
@@ -113,80 +119,131 @@ Neighborlist<RealType>::get_nblist_host(int N, const double *h_coords, const dou
+    this->compute_block_bounds_device(N, K, D, d_col_coords, d_row_coords, d_box, stream);
+
+    const int tpb = 32;
+    const int column_blocks = this->column_blocks();
+    const int row_blocks = this->B();
+    const int Y = this->Y();
 
-    int tpb = 32;
-    const int B = this->B(); // (N+32-1)/32;
-    const int Y = this->Y(); // (B+32-1)/32;
+    dim3 dimGrid(row_blocks, Y, 1); // block x, y, z dims
 
-    dim3 dimGrid(B, Y, 1); // block x, y, z dims
+    const bool upper_triangular = K_ == 0;
 
     // (ytz): TBD shared memory, stream
-    k_find_blocks_with_ixns<RealType><<<dimGrid, tpb, 0, stream>>>(
-        N,
-        d_block_bounds_ctr_,
-        d_block_bounds_ext_,
-        d_coords,
-        d_box,
-        d_ixn_count_,
-        d_ixn_tiles_,
-        d_ixn_atoms_,
-        d_trim_atoms_,
-        cutoff);
+    if (upper_triangular) {
+        // Compute only the upper triangle as rows and cols are the same
+        // pass duplicates of column coords and the bounding boxes
+        k_find_blocks_with_ixns<RealType, true><<<dimGrid, tpb, 0, stream>>>(
+            N,
+            N,
+            d_col_block_bounds_ctr_,
+            d_col_block_bounds_ext_,
+            d_col_block_bounds_ctr_,
+            d_col_block_bounds_ext_,
+            d_col_coords,
+            d_col_coords,
+            d_box,
+            d_ixn_count_,
+            d_ixn_tiles_,
+            d_ixn_atoms_,
+            d_trim_atoms_,
+            cutoff);
+    } else {
+        k_find_blocks_with_ixns<RealType, false><<<dimGrid, tpb, 0, stream>>>(
+            N,
+            K,
+            d_col_block_bounds_ctr_,
+            d_col_block_bounds_ext_,
+            d_row_block_bounds_ctr_,
+            d_row_block_bounds_ext_,
+            d_col_coords,
+            d_row_coords,
+            d_box,
+            d_ixn_count_,
+            d_ixn_tiles_,
+            d_ixn_atoms_,
+            d_trim_atoms_,
+            cutoff);
+    }
 
     gpuErrchk(cudaPeekAtLastError());
-
-    k_compact_trim_atoms<<<B, tpb, 0, stream>>>(N, Y, d_trim_atoms_, d_ixn_count_, d_ixn_tiles_, d_ixn_atoms_);
+    k_compact_trim_atoms<<<row_blocks, tpb, 0, stream>>>(N, Y, d_trim_atoms_, d_ixn_count_, d_ixn_tiles_, d_ixn_atoms_);
 
     gpuErrchk(cudaPeekAtLastError());
 }
 
+template <typename RealType>
+void Neighborlist<RealType>::build_nblist_device(
+    const int N, const double *d_coords, const double *d_box, const double cutoff, cudaStream_t stream) {
+
+    this->build_nblist_device(N, 0, d_coords, nullptr, d_box, cutoff, stream);
+}
+
 template <typename RealType>
 void Neighborlist<RealType>::compute_block_bounds_device(
-    const int N,            // Number of atoms
-    const int D,            // Box dimensions
-    const double *d_coords, // [N*3]
-    const double *d_box,    // [D*3]
+    const int N,                // Number of atoms in column
+    const int K,                // Number of atoms in row
```

Comment:
This naming feels a bit strange to me, since I'm used to `N` being the total number. Is this conventional, or could it make sense to give a different name to the number of atoms in a column? (it's possible this will make more sense to me once I have a better understanding of what's going on here)

Reply:
I find that N is just a size of 'something', I can change them more broadly